scanner/src/main/kotlin/scanners/fossid/FossId.kt

```diff
@@ -0,0 +1,396 @@
+                false
+            } else {
+                when (response.data?.state) {
+                    ScanState.FINISHED -> {
+                        this@FossId.log.info { "Scan result: ${response.data?.comment}." }
+                        true
+                    }
+                    else -> {
+                        this@FossId.log.info {
+                            "Scan status for '$scanCode' is '${response.data?.state}'. " +
+                                    "Waiting..."
+                        }
+                        false
+                    }
+                }
+            }
+        }
+        requireNotNull(timeout) {
+            "Timeout while waiting for the scan to complete"
+        }
+        log.info {
+            "Scan has been completed."
+        }
+    }
+
+    /**
+     * Get the different kind of results from the scan with [scanCode]
+     */
+    private suspend fun getRawResults(scanCode: String): RawResults {
+        val responseIdentifiedFiles = service.listIdentifiedFiles(user, apiKey, scanCode)
+            .checkResponse("list identified files", false)
+
+        // TODO Replace the return type of listIdentifiedFiles to Call<EntityPostResponseBody<Any>> and use toList().
+        val identifiedFiles = responseIdentifiedFiles.data?.values?.toList().orEmpty()
+        log.info("${identifiedFiles.size} identified files have been returned for scan_code=$scanCode.")
+
+        val responseMarkedAsIdentified = service.listMarkedAsIdentifiedFiles(user, apiKey, scanCode)
+            .checkResponse("list marked as identified files", false)
+        requireNotNull(responseMarkedAsIdentified)
+        val markedAsIdentifiedFiles = responseMarkedAsIdentified.toList(MarkedAsIdentifiedFile::class)
+        log.info(
+            "${markedAsIdentifiedFiles.size} marked as identified files " +
+                    "have been returned for scan_code=$scanCode."
+        )
+
+        // The match_type=ignore info is already in scanResults, but here we also get the ignore reason.
+        val responseListIgnoredFiles = service.listIgnoredFiles(user, apiKey, scanCode)
+            .checkResponse("list ignored files", false)
+        val listIgnoredFiles = responseListIgnoredFiles.toList(IgnoredFile::class)
+        return RawResults(identifiedFiles, markedAsIdentifiedFiles, listIgnoredFiles)
+    }
+
+    /**
+     * Construct the [ScanSummary] for this FossId scan.
+     */
+    private fun createResultSummary(startTime: Instant, rawResults: RawResults): ScanResult {
+        val associate = rawResults.listIgnoredFiles.associateBy { it.path }
+
+        val (filesCount, licenseFindings, copyrightFindings) = kotlin.run {
+            if (rawResults.markedAsIdentifiedFiles.isNotEmpty()) {
+                rawResults.markedAsIdentifiedFiles.mapSummary(associate)
+            } else {
+                rawResults.identifiedFiles.mapSummary(associate)
+            }
+        }
```

Comment:
val (filesCount, licenseFindings, copyrightFindings) = if (rawResults.markedAsIdentifiedFiles.isNotEmpty()) {
            rawResults.markedAsIdentifiedFiles
        } else {
            rawResults.identifiedFiles
        }.mapSummary(associate)

Reply:
Even better, I inverted the test to avoid the negation in the test condition with `isEmpty`.